core/store/src/trie/update.rs

```diff
@@ -79,32 +87,22 @@ impl TrieUpdate {
         &self.committed
     }
 
-    pub fn set(&mut self, key: TrieKey, value: Vec<u8>) {
-        self.prospective.insert(key.to_vec(), Some(value));
+    pub fn set(&mut self, trie_key: TrieKey, value: Vec<u8>) {
+        self.prospective
+            .insert(trie_key.to_vec(), TrieKeyValueUpdate { trie_key, value: Some(value) });
```

Comment:
This seems like an unnecessary duplication. We store trie_key twice for each record -- once in deserialized form and another in non-deserialized. I don't see a reason why we cannot store it in non-deserialized form here in keys of BTreeMap.

Reply:
@evgenykuzyakov If we want to avoid the duplication, we will need to keep the key mangling I suggested to kill in the comment above...

Reply:
It's to avoid `Vec<u8>` construction every time `TrieKey`. We'll have to cache it within `TrieKey`. Also we later use `Vec<u8>` to write it into the Trie, while TrieKey are needed for `RawStateChangesWithTrieKey`.
If we do a little bit of refactoring, then we can avoid cloning them in `Runtime::apply`. So finalize would actually use different record.
Created an issue for this #2327

Reply:
@evgenykuzyakov I am lost with your explanation. I think Max refers to the need to include `trie_key` in `TrieKeyValueUpdate` while the same key is used for the RocksDB key, so we duplicate the information in our storage.

Reply:
I think we need `Vec<u8>` for `Trie` itself, but also need it for `BTreeMap` and for iterators. If we don't cache it within `TrieKey`, then `BTreeMap` performance be worse.

But we also want to keep `TrieKey` for `RawStateChangesWithTrieKey` in the linked issue.

Reply:
Please add comment with the issue number on this line too.